internal/pkg/cli/app_deploy.go

```diff
@@ -362,79 +358,106 @@ func (o *appDeployOpts) getAppDockerfilePath() (string, error) {
 	return strings.TrimSuffix(mf.DockerfilePath(), "/Dockerfile"), nil
 }
 
-func (o *appDeployOpts) retrieveTemplate() (*cfnTemplates, error) {
-	appBuffer := &bytes.Buffer{}
-	addonsBuffer := &bytes.Buffer{}
-
-	appPackage := packageAppOpts{
-		packageAppVars: packageAppVars{
-			AppName:    o.AppName,
-			EnvName:    o.targetEnvironment.Name,
-			Tag:        o.ImageTag,
-			GlobalOpts: o.GlobalOpts,
-		},
-
-		stackWriter:   appBuffer,
-		paramsWriter:  ioutil.Discard,
-		addonsWriter:  addonsBuffer,
-		initAddonsSvc: initPackageAddonsSvc,
-		store:         o.projectService,
-		describer:     o.appPackageCfClient,
-		ws:            o.workspaceService,
-	}
-
-	if err := appPackage.Execute(); err != nil {
-		return nil, fmt.Errorf("retrieve template: %w", err)
-	}
-	return &cfnTemplates{
-		addons: addonsBuffer,
-		app:    appBuffer,
-	}, nil
-}
-
-func (o *appDeployOpts) pushAddonsTemplateToS3Bucket(addonsTemplate *bytes.Buffer) (string, error) {
-	if addonsTemplate.String() == "" {
-		return "", nil
+// pushAddonsTemplateToS3Bucket generates the addons template for the application and pushes it to S3.
+// If the application doesn't have any addons, it returns the empty string and no errors.
+// If the application has addons, it returns the URL of the S3 object storing the addons template.
+func (o *appDeployOpts) pushAddonsTemplateToS3Bucket() (string, error) {
+	template, err := o.addonsSvc.Template()
+	if err != nil {
+		var notExistErr *addons.ErrDirNotExist
+		if errors.As(err, &notExistErr) {
+			// addons doesn't exist for app, the url is empty.
+			return "", nil
+		}
+		return "", fmt.Errorf("retrieve addons template: %w", err)
 	}
-	resources, err := o.appPackageCfClient.GetProjectResourcesByRegion(o.targetProject, o.targetEnvironment.Region)
+	resources, err := o.projectCFSvc.GetProjectResourcesByRegion(o.targetProject, o.targetEnvironment.Region)
 	if err != nil {
 		return "", fmt.Errorf("get project resources: %w", err)
 	}
 
-	url, err := o.s3Service.PutArtifact(resources.S3Bucket, fmt.Sprintf(archer.AddonsCfnTemplateNameFormat, o.AppName), addonsTemplate)
+	reader := strings.NewReader(template)
+	url, err := o.s3Service.PutArtifact(resources.S3Bucket, fmt.Sprintf(archer.AddonsCfnTemplateNameFormat, o.AppName), reader)
 	if err != nil {
 		return "", fmt.Errorf("put addons artifact to bucket %s: %w", resources.S3Bucket, err)
 	}
 	return url, nil
 }
 
-func (o *appDeployOpts) deployAppStack(appTemplate *bytes.Buffer, addonsURL string) error {
-	id, err := uuid.NewRandom()
+func (o *appDeployOpts) manifest() (interface{}, error) {
```

Comment:
Is this function only to get the type of the application 🤔feel like we can store the app type in param store.

Reply:
it's fetching the whole manifest object not just the type :) the whole object is needed for the `stack.NewLoadBalancedWebApp` constructor